# $$Formation\ de\ Renforcement\ du\ Secteur\ Technologique$$
# $$(FRST/FDS)$$

## $$Cours\ d'Initiation\ au\ Machine\ Learning$$
### $$Apprentissage\ Non-Supervisé$$

### Professeur : *M. Evens TOUSSAINT*

### Apprenant : *Lionel PETIOTE*

## $$TP\ Final\ ML - NS$$

![Unsupervised Learning](https://tse2.mm.bing.net/th/id/OIP.CD0lIEEEPFWqMRoLIWh0XAHaEK?rs=1&pid=ImgDetMain&o=7&rm=3)

# IMPORTATIONS

In [3]:
# Importation des bibliothèques
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

# Bibliothèques de visualisation
import matplotlib.pyplot as plt
import seaborn as sns

# Gestion d'erreurs
import warnings
warnings.filterwarnings('ignore')

# TRAVAIL DEMANDE

## 1- Chargement et présentation des données

In [4]:
# Chargement des données
from sklearn.datasets import load_wine
data = load_wine()
df = pd.DataFrame(data.data, columns=data.feature_names)

# Présentation des données
print(f"Le dataset est composé de {df.shape[0]} observations et de {df.shape[1]} variables.")
print('\nAperçu des données :')
display(df.head())
print('\nRésumé informationnel :\n')
df.info()
print('\nStatistiques descriptives :')
df.describe()

Le dataset est composé de 178 observations et de 13 variables.

Aperçu des données :


,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline
0,14.23,1.71,2.43,15.6,127.0,2.80,3.06,0.28,2.29,5.64,1.04,3.92,1065.0
1,13.20,1.78,2.14,11.2,100.0,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050.0
2,13.16,2.36,2.67,18.6,101.0,2.80,3.24,0.30,2.81,5.68,1.03,3.17,1185.0
3,14.37,1.95,2.50,16.8,113.0,3.85,3.49,0.24,2.18,7.80,0.86,3.45,1480.0
4,13.24,2.59,2.87,21.0,118.0,2.80,2.69,0.39,1.82,4.32,1.04,2.93,735.0



Résumé informationnel :

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline
count,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000
mean,13.000618,2.336348,2.366517,19.494944,99.741573,2.295112,2.029270,0.361854,1.590899,5.058090,0.957449,2.611685,746.893258
std,0.811827,1.117146,0.274344,3.339564,14.282484,0.625851,0.998859,0.124453,0.572359,2.318286,0.228572,0.709990,314.907474
min,11.030000,0.740000,1.360000,10.600000,70.000000,0.980000,0.340000,0.130000,0.410000,1.280000,0.480000,1.270000,278.000000
25%,12.362500,1.602500,2.210000,17.200000,88.000000,1.742500,1.205000,0.270000,1.250000,3.220000,0.782500,1.937500,500.500000
50%,13.050000,1.865000,2.360000,19.500000,98.000000,2.355000,2.135000,0.340000,1.555000,4.690000,0.965000,2.780000,673.500000
75%,13.677500,3.082500,2.557500,21.500000,107.000000,2.800000,2.875000,0.437500,1.950000,6.200000,1.120000,3.170000,985.000000
max,14.830000,5.800000,3.230000,30.000000,162.000000,3.880000,5.080000,0.660000,3.580000,13.000000,1.710000,4.000000,1680.000000


### $Commentaire$

##### Le jeu de données est composé de 178 enregistrements caractérisés par 13 variables, toutes de type numérique.
- Pas de valeur manquante

##### Les variables présentent des ordres de grandeur très différents. Par exemple :

- Proline peut atteindre des valeurs supérieures à 1000
- Magnesium se situe autour de 100
- Hue varie généralement entre 0 et 2

## 2- Prétraitement et standardisation

##### Une analyse des variables du jeu de données montre une forte disparité dans l’échelle de distribution, ce qui implique que certaines variables ont une variance beaucoup plus élevée que d’autres.
##### Or, les méthodes qu'on aura à utiliser dans cette étude, comme K-Means, DBSCAN ou le clustering agglomératif, reposent sur des mesures de distance (notamment la distance euclidienne).

##### Sans standardisation :
- Les variables à grande échelle (comme Proline) domineraient le calcul des distances
- Les variables à petite échelle (comme Hue) auraient un impact négligeable
- Cela introduirait un biais dans la formation des clusters

##### La standardisation permet :
- d’éliminer l’effet des différences d’échelle entre variables
- d’assurer une contribution équitable de chaque variable
- d’obtenir des résultats de clustering plus fiables et interprétables

In [ ]:
# Standardisation
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df)

## 3- Application de K-Means

In [ ]:
# K-Means
inertia = []
for k in range(1, 10):
    km = KMeans(n_clusters=k, random_state=42)
    km.fit(X_scaled)
    inertia.append(km.inertia_)